x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/merge_unwired_stream_pipeline_actions.ts

```diff
@@ -0,0 +1,338 @@
+  DeleteProcessorFromIngestPipelineAction,
+  ElasticsearchAction,
+  UpsertWriteIndexOrRolloverAction,
+} from './execution_plan';
+
+const MANAGED_BY_STREAMS = 'streams';
+
+type UnwiredStreamPipelineAction =
+  | AppendProcessorToIngestPipelineAction
+  | DeleteProcessorFromIngestPipelineAction;
+
+// Should this be a static method on UnwiredStream instead?
+export async function mergeUnwiredStreamPipelineActions(
+  actionsByType: ActionsByType,
+  scopedClusterClient: IScopedClusterClient
+) {
+  const maybeActions = [
+    ...actionsByType.append_processor_to_ingest_pipeline,
+    ...actionsByType.delete_processor_from_ingest_pipeline,
+  ] as UnwiredStreamPipelineAction[];
+
+  if (maybeActions.length === 0) {
+    return;
+  }
+
+  const actionsByPipeline = groupBy(maybeActions, 'targetPipeline');
+
+  for (const [pipelineName, actions] of Object.entries(actionsByPipeline)) {
+    const pipeline = await getPipeline(pipelineName, scopedClusterClient);
+
+    if (pipeline) {
+      if (pipeline._meta?.managed_by === MANAGED_BY_STREAMS) {
+        updateExistingStreamsManagedPipeline({
+          pipelineName,
+          pipeline,
+          actions,
+          actionsByType,
+          scopedClusterClient,
+        });
+      } else {
+        updateExistingUserManagedPipeline({
+          pipelineName,
+          actions,
+          actionsByType,
+          scopedClusterClient,
+        });
+      }
+    } else {
+      createStreamsManagedPipeline({
+        actions,
+        actionsByType,
+        scopedClusterClient,
+      });
+    }
+  }
+
+  actionsByType.append_processor_to_ingest_pipeline = [];
+  actionsByType.delete_processor_from_ingest_pipeline = [];
+}
+
+async function createStreamsManagedPipeline({
+  actions,
+  actionsByType,
+  scopedClusterClient,
+}: {
+  actions: UnwiredStreamPipelineAction[];
+  actionsByType: ActionsByType;
+  scopedClusterClient: IScopedClusterClient;
+}) {
+  assertOnlyAppendActions(actions);
+  const targetTemplateNames = uniq(actions.map((action) => action.template));
+  if (targetTemplateNames.length !== 1) {
+    throw new Error('Append actions targeting the same new pipeline target different templates');
+  }
+  const indexTemplate = await getIndexTemplate(targetTemplateNames[0], scopedClusterClient);
+
+  const pipelineName = `${indexTemplate.name}@custom`;
+
+  actionsByType.upsert_ingest_pipeline.push({
+    type: 'upsert_ingest_pipeline',
+    // All of these are UnwiredStreams so take any stream name to use for the ordering of operations
+    stream: actions[0].dataStream,
+    request: {
+      id: pipelineName,
+      processors: actions.map((action) => action.processor),
+      _meta: {
+        description: `Streams managed pipeline to connect Unwired streams to the Streams layer`,
+        managed: true,
+        managed_by: MANAGED_BY_STREAMS,
+      },
+      version: ASSET_VERSION,
+    },
+  });
+
+  actionsByType.upsert_index_template.push({
```

Comment:
Why did you pull the logic to upsert the index template, finding the right pipeline to modify and so on in here? I know it's just about where to put the logic, but to me the UnwiredStream class was a good place for these.

I thought what we said is that basically the only logic that needs to go here is to assemble the full pipeline out of its `append_processor_to_ingest_pipeline` and `delete_processor_from_ingest_pipeline` actions - the rest can stay together with the other logic.

Reply:
After syncing on Zoom, we reached the conclusion that in order to clean up all Streams created resources, we need to pull this logic out of the `UnwiredStream` so that we can decide **when** to delete the Streams managed ingest pipeline and update the index template to remove the `default_pipeline` setting.

If we try to do actions only per stream, then we can never decide to do this clean up, which would mean that all `UnwiredStreams` may be deleted but the index template still points to the Streams managed ingest pipeline (which also still exists but is empty). 

While it would be nice to keep as much logic inside of the `StreamsActiveRecord` classes we felt it was important to be able to clean up properly.